File: CS/GrowthWare.Framework/Model/Profiles/MAccountProfile.cs
using GrowthWare.Framework.Model.Profiles.Base;
using GrowthWare.Framework.Model.Profiles.Interfaces;
using System;
using System.Collections;
using System.Collections.ObjectModel;
using System.Data;


namespace GrowthWare.Framework.Model.Profiles
{
    /// <summary>
    /// Properties for an account.
    /// </summary>
    [Serializable(), CLSCompliant(true)]
    public class MAccountProfile : MProfile, IMGroupRoleSecurity
    {

        #region "Member fields"
        private Collection<string> m_AssignedRoles = new Collection<string>();
        private Collection<string> m_AssignedGroups = new Collection<string>();
        private Collection<string> m_DerivedRoles = new Collection<string>();
        private static String s_RoleColumn = "Roles";
        private static String s_GroupColumn = "Groups";
        #endregion

        #region Private Methods
        /// <summary>
        /// Sets the assigned roles or groups.
        /// </summary>
        /// <param name="StringCollectionObject">The collection of roles or groups that need to be set</param>
        /// <param name="GroupsOrRoles">The DataRowCollection that represents either roles or groups</param>
        /// <param name="ColumnName">The column name to retrieve the data from</param>
        private static void setRolesOrGroups(ref Collection<string> StringCollectionObject, DataRowCollection GroupsOrRoles, String ColumnName)
        {
            foreach (DataRow mRow in GroupsOrRoles)
            {
                if (!Convert.IsDBNull(mRow[ColumnName]))
                {
                    StringCollectionObject.Add(mRow[ColumnName].ToString());
                }
            }
        }

        /// <summary>
        /// Sets the assigned roles or groups.
        /// </summary>
        /// <param name="StringCollectionObject">The collection of roles or groups that need to be set</param>
        /// <param name="CommaSeparatedString">A comma Separated list of roles or groups 'you, me' as an example</param>
        private static void setRolesOrGroups(ref Collection<String> StringCollectionObject, ref String CommaSeparatedString)
        {
            StringCollectionObject = new Collection<string>();
            string[] mRoles = CommaSeparatedString.Split(',');
            foreach (var mRole in mRoles)
            {
                StringCollectionObject.Add(mRole.ToString());
            }
        }

        private static string getCommaSeportatedString(Collection<string> CollectionOfStrings)
        {
            String mRetValue = String.Empty;
            if (CollectionOfStrings != null)
            {
                if (CollectionOfStrings.Count > 0)
                {
                    foreach (var item in CollectionOfStrings)
                    {
                        mRetValue += item.ToString() + ",";
                    }
                }
            }
            if (mRetValue.Length > 0)
            {
                mRetValue = mRetValue.Substring(0, mRetValue.Length - 1);
            }
            return mRetValue;
        }
        #endregion

        #region "Protected Methods"
        /// <summary>
        /// Populates direct properties as well as passing the DataRow to the abstract class
        /// for the population of the base properties.
        /// </summary>
        /// <param name="dataRow">DataRow</param>
        protected new void Initialize(DataRow dataRow)
        {
            base.NameColumnName = "ACCT";
            base.IdColumnName = "ACCT_SEQ_ID";
            base.Initialize(dataRow);
            Account = base.Name;
            this.Email = base.GetString(dataRow, "EMAIL");
            this.EnableNotifications = base.GetBool(dataRow, "ENABLE_NOTIFICATIONS");
            this.IsSystemAdmin = base.GetBool(dataRow, "IS_SYSTEM_ADMIN");
            this.Status = base.GetInt(dataRow, "STATUS_SEQ_ID");
            this.Password = base.GetString(dataRow, "PWD");
            this.FailedAttempts = base.GetInt(dataRow, "FAILED_ATTEMPTS");
            this.FirstName = base.GetString(dataRow, "FIRST_NAME");
            this.LastLogOn = base.GetDateTime(dataRow, "LAST_LOGIN", DateTime.Now);
            this.LastName = base.GetString(dataRow, "LAST_NAME");
            this.Location = base.GetString(dataRow, "LOCATION");
            this.PasswordLastSet = base.GetDateTime(dataRow, "PASSWORD_LAST_SET", DateTime.Now);
            this.MiddleName = base.GetString(dataRow, "MIDDLE_NAME");
            this.PreferredName = base.GetString(dataRow, "PREFERRED_NAME");
            this.TimeZone = base.GetInt(dataRow, "TIME_ZONE");
        }
        #endregion

        #region "Public Methods"

        /// <summary>
        /// Provides a new account profile with the default vaules
        /// </summary>
        /// <remarks></remarks>
        public MAccountProfile()
        {
        }

        /// <summary>
        /// Will populate values based on the contents of the data row.
        /// </summary>
        /// <param name="detailRow">Datarow containing base values</param>
        /// <remarks>
        /// Class should be inherited to extend to your project specific properties
        /// </remarks>
        public MAccountProfile(DataRow detailRow)
        {
            this.Initialize(detailRow);
        }

        /// <summary>
        /// Will populate values based on the contents of the data row.
        /// Also populates the roles and gropus properties.
        /// </summary>
        /// <param name="detailRow">DataRow containing base values</param>
        /// <param name="assignedRolesData">DataRow containing Role data</param>
        /// <param name="assignedGroupsData">DataRow containing Group data</param>
        /// <param name="derivedRolesData">DataRow containing Role data derived from both assigned roles and groups.></param>
        /// <remarks>
        /// Class should be inherited to extend to your project specific properties
        /// </remarks>
        public MAccountProfile(DataRow detailRow, DataTable assignedRolesData, DataTable assignedGroupsData, DataTable derivedRolesData)
        {
            if (detailRow != null) 
            {
                this.Initialize(detailRow);
                if (assignedRolesData != null) setRolesOrGroups(ref m_AssignedRoles, assignedRolesData.Rows, s_RoleColumn);
                if (assignedGroupsData != null) setRolesOrGroups(ref m_AssignedGroups, assignedGroupsData.Rows, s_GroupColumn);
                if (derivedRolesData != null) setRolesOrGroups(ref m_DerivedRoles, derivedRolesData.Rows, s_RoleColumn);
            }
        }

        /// <summary>
        /// Will set the collection of roles given a comma Separated string of roles.
        /// </summary>
        /// <param name="commaSeparatedRoles">String of comma Separated roles</param>
        public void SetRoles(string commaSeparatedRoles)
        {
            setRolesOrGroups(ref m_AssignedRoles, ref commaSeparatedRoles);
        }

        /// <summary>
        /// Will set the collection of groups given a comma Separated string of groups.
        /// </summary>
        /// <param name="commaSeparatedGroups">String of comma Separated groups</param>
        public void SetGroups(string commaSeparatedGroups)
        {
            setRolesOrGroups(ref m_AssignedGroups, ref commaSeparatedGroups);
        }

        #endregion

        #region "Public Properties"

        /// <summary>
        /// Represents the roles that have been directly assigned to the account.
        /// </summary>
        public Collection<String> AssignedRoles
        {
            get
            {
                return m_AssignedRoles;
            }
        }

        /// <summary>
        /// Represents the groups that have been directly assigned to the account.
        /// </summary>
        public Collection<String> Groups
        {
            get
            {
                return m_AssignedGroups;
            }
        }

        /// <summary>
        /// Represents the roles that have been assigned either directly or through assoication of a role to a group.
        /// </summary>
        public Collection<String> DerivedRoles
        {
            get
            {
                return m_DerivedRoles;
            }
        }

        /// <summary>
        /// Represents the account
        /// </summary>
        public String Account { get; set; }

        /// <summary>
        /// Represents the email address
        /// </summary>
        public String Email { get; set; }

        /// <summary>
        /// Used to determine if the client would like to recieve notifications.
        /// </summary>
        public bool EnableNotifications { get; set; }

        /// <summary>
        /// Represents the status of the account
        /// </summary>
        public int Status { get; set; }

        /// <summary>
        /// Indicates the last time the account password was changed
        /// </summary>
        public DateTime PasswordLastSet { get; set; }

        /// <summary>
        /// The password for the account
        /// </summary>
        public String Password { get; set; }

        /// <summary>
        /// The number of failed logon attemps
        /// </summary>
        public int FailedAttempts { get; set; }

        /// <summary>
        /// First name of the person for the account
        /// </summary>
        public String FirstName { get; set; }

        /// <summary>
        /// Indicates if the account is a system administrator ... used to
        /// prevent complete lockout when the roles have been
        /// damaged.
        /// </summary>
        public bool IsSystemAdmin { get; set; }

        /// <summary>
        /// Last name of the person for the account
        /// </summary>
        public String LastName { get; set; }

        /// <summary>
        /// Middle name of the person for the account
        /// </summary>
        public String MiddleName { get; set; }

        /// <summary>
        /// Prefered or nick name of the person for the account
        /// </summary>
        public String PreferredName { get; set; }

        /// <summary>
        /// The timezone for the account
        /// </summary>
        public int TimeZone { get; set; }

        /// <summary>
        /// The location of the account
        /// </summary>
        public string Location { get; set; }

        /// <summary>
        /// The date and time the account was last loged on
        /// </summary>
        public DateTime LastLogOn { get; set; }

        /// <summary>
        /// Converts the collection of AssignedRoles to a comma Separated string.
        /// </summary>
        /// <returns>String</returns>
        public String GetCommaSeparatedAssignedRoles
        {
            get { return getCommaSeportatedString(m_AssignedRoles); }
        }

        /// <summary>
        /// Converts the collection of AssignedGroups to a comma Separated string.
        /// </summary>
        /// <returns>String</returns>
        public String GetCommaSeparatedAssignedGroups
        {
            get { return getCommaSeportatedString(m_AssignedGroups); }
        }

        /// <summary>
        /// Converts the collection of DerivedRoles to a comma Separated string.
        /// </summary>
        /// <returns>String</returns>
        public String GetCommaSeparatedDerivedRoles
        {
            get { return getCommaSeportatedString(m_DerivedRoles); }
        }
        #endregion

    }
}

File: VB/GrowthWare.Framework/Model/Profiles/MAccountProfile.vb
Imports GrowthWare.Framework.Model.Profiles.Base
Imports System.Collections.ObjectModel
Imports GrowthWare.Framework.Model.Profiles.Interfaces

Namespace Model.Profiles
    ''' <summary>
    ''' Base properties an account Profile you can inherit this class and add
    ''' any properties or methods that suit your project needs.
    ''' </summary>
    ''' <remarks>
    ''' Corresponds to table [ZGWSecurity].[Accounts] and 
    ''' Store procedures: 
    ''' [ZGWSecurity].[Set_Account], [ZGWSecurity].[Get_Account]
    ''' </remarks>
    <Serializable(), CLSCompliant(True)> _
    Public Class MAccountProfile
        Inherits MProfile
        Implements IMGroupRoleSecurity

#Region "Constructors"
        ''' <summary>
        ''' Provides a new account profile with the default vaules
        ''' </summary>
        ''' <remarks></remarks>
        Public Sub New()

        End Sub

        ''' <summary>
        ''' Will populate values based on the contents of the data row.
        ''' </summary>
        ''' <param name="detailRow">Datarow containing base values</param>
        ''' <remarks>
        ''' Class should be inherited to extend to your project specific properties
        ''' </remarks>
        Public Sub New(ByVal detailRow As DataRow)
            Me.IdColumnName = "ACCT_SEQ_ID"
            Me.NameColumnName = "ACCT"
            Me.Initialize(detailRow)
        End Sub

        Public Sub New(ByVal detailRow As DataRow, ByVal assignedRolesData As DataTable, ByVal assignedGroupsData As DataTable, ByVal derivedRolesData As DataTable)
            If Not detailRow Is Nothing Then
                Me.Initialize(detailRow)
                If Not assignedRolesData Is Nothing Then setRolesOrGroups(m_AssignedRoles, assignedRolesData.Rows, m_RoleColumn)
                If Not assignedGroupsData Is Nothing Then setRolesOrGroups(m_Groups, assignedGroupsData.Rows, m_GroupColumn)
                If Not derivedRolesData Is Nothing Then setRolesOrGroups(m_DerivedRoles, derivedRolesData.Rows, m_RoleColumn)
            End If
        End Sub
#End Region

#Region "Member Fields"
        Private m_RoleColumn As String = "Roles"
        Private m_GroupColumn As String = "Groups"
        Private m_AssignedRoles As Collection(Of String) = New Collection(Of String)
        Private m_Groups As Collection(Of String) = New Collection(Of String)
        Private m_DerivedRoles As Collection(Of String) = New Collection(Of String)
#End Region

#Region "Public Properties"

        ''' <summary>
        ''' Represents the roles that have been directly assigned to the account.
        ''' </summary>
        Public ReadOnly Property AssignedRoles As Collection(Of String) Implements IMGroupRoleSecurity.AssignedRoles
            Get
                Return m_AssignedRoles
            End Get
        End Property

        ''' <summary>
        ''' Represents the roles that have been assigned either directly or through assoication of a role to a group.
        ''' </summary>
        Public ReadOnly Property DerivedRoles As Collection(Of String) Implements IMGroupRoleSecurity.DerivedRoles
            Get
                Return m_DerivedRoles
            End Get
        End Property

        ''' <summary>
        ''' Represents the groups that have been directly assigned to the account.
        ''' </summary>
        Public ReadOnly Property Groups As Collection(Of String) Implements IMGroupRoleSecurity.Groups
            Get
                Return m_Groups
            End Get
        End Property

        ''' <summary>
        ''' Represents the account
        ''' </summary>
        Public Property Account As String

        ''' <summary>
        ''' Represents the email address
        ''' </summary>
        Public Property Email() As String

        ''' <summary>
        ''' Represents the status of the account
        ''' </summary>
        Public Property Status() As Integer

        ''' <summary>
        ''' Indicates the last time the account password was changed
        ''' </summary>
        Public Property PasswordLastSet() As DateTime

        ''' <summary>
        ''' The password for the account
        ''' </summary>
        Public Property Password() As String

        ''' <summary>
        ''' The number of failed logon attemps
        ''' </summary>
        Public Property FailedAttempts() As Integer

        ''' <summary>
        ''' First name of the person for the account
        ''' </summary>
        Public Property FirstName() As String

        ''' <summary>
        ''' Indicates if the account is a system administrator ... used to
        ''' prevent complete lockout when the roles have been
        ''' damaged.
        ''' </summary>
        Public Property IsSystemAdmin() As Boolean

        ''' <summary>
        ''' Last name of the person for the account
        ''' </summary>
        Public Property LastName() As String

        ''' <summary>
        ''' Middle name of the person for the account
        ''' </summary>
        Public Property MiddleName() As String

        ''' <summary>
        ''' Prefered or nick name of the person for the account
        ''' </summary>
        Public Property PreferredName() As String

        ''' <summary>
        ''' The timezone for the account
        ''' </summary>
        Public Property TimeZone() As Integer

        ''' <summary>
        ''' The location of the account
        ''' </summary>
        Public Property Location() As String

        ''' <summary>
        ''' The date and time the account was last loged on
        ''' </summary>
        Public Property LastLogOn() As DateTime

        ''' <summary>
        ''' Used to determine if the client would like to recieve notifications.
        ''' </summary>
        Public Property EnableNotifications() As Boolean
        ''' <summary>
        ''' Converts the collection of AssignedRoles to a comma Separated string.
        ''' </summary>
        ''' <returns>String</returns>
        Public ReadOnly Property GetCommaSeparatedAssignedRoles As String
            Get
                Return getCommaSeportatedString(m_AssignedRoles)
            End Get
        End Property


        ''' <summary>
        ''' Converts the collection of AssignedGroups to a comma Separated string.
        ''' </summary>
        ''' <returns>String</returns>
        Public ReadOnly Property GetCommaSeparatedAssignedGroups As String
            Get
                Return getCommaSeportatedString(m_Groups)
            End Get
        End Property

        ''' <summary>
        ''' Converts the collection of DerivedRoles to a comma Separated string.
        ''' </summary>
        ''' <returns>String</returns>
        Public ReadOnly Property GetCommaSeparatedDerivedRoles As String
            Get
                Return getCommaSeportatedString(m_DerivedRoles)
            End Get
        End Property

#End Region

#Region "Protected Methods"
        ''' <summary>
        ''' Populates direct properties as well as passing the DataRow to the abstract class
        ''' for the population of the base properties.
        ''' </summary>
        ''' <param name="DetailRow">DataRow</param>
        Protected Shadows Sub Initialize(ByVal detailRow As DataRow)
            MyBase.IdColumnName = "Account_SeqID"
            MyBase.NameColumnName = "ACCT"
            MyBase.Initialize(detailRow)
            Account = MyBase.GetString(detailRow, "ACCT")
            Email = MyBase.GetString(detailRow, "EMAIL")
            EnableNotifications = MyBase.GetBool(detailRow, "ENABLE_NOTIFICATIONS")
            IsSystemAdmin = MyBase.GetBool(detailRow, "IS_SYSTEM_ADMIN")
            Status = MyBase.GetInt(detailRow, "STATUS_SEQ_ID")
            Password = MyBase.GetString(detailRow, "PWD")
            FailedAttempts = MyBase.GetInt(detailRow, "FAILED_ATTEMPTS")
            FirstName = MyBase.GetString(detailRow, "FIRST_NAME")
            LastLogOn = MyBase.GetDateTime(detailRow, "LAST_LOGIN", Date.Now)
            LastName = MyBase.GetString(detailRow, "LAST_NAME")
            Location = MyBase.GetString(detailRow, "LOCATION")
            PasswordLastSet = MyBase.GetDateTime(detailRow, "PASSWORD_LAST_SET", Date.Now)
            MiddleName = MyBase.GetString(detailRow, "MIDDLE_NAME")
            PreferredName = MyBase.GetString(detailRow, "Preferred_Name")
            TimeZone = MyBase.GetInt(detailRow, "TIME_ZONE")
        End Sub
#End Region

#Region "Public Methods"

        ''' <summary>
        ''' Will set the collection of roles given a comma Separated string of roles.
        ''' </summary>
        ''' <param name="commaSeparatedRoles">String of comma Separated roles</param>
        Public Sub SetRoles(ByVal commaSeparatedRoles As String)
            setRolesOrGroups(m_AssignedRoles, commaSeparatedRoles)
        End Sub

        ''' <summary>
        ''' Will set the collection of groups given a comma Separated string of groups.
        ''' </summary>
        ''' <param name="commaSeparatedGroups">String of comma Separated groups</param>
        Public Sub SetGroups(ByVal commaSeparatedGroups As String)
            setRolesOrGroups(m_Groups, commaSeparatedGroups)
        End Sub
#End Region

#Region "Private Methods"
        ''' <summary>
        ''' Sets the assigned roles or groups.
        ''' </summary>
        ''' <param name="StringCollectionObject">The collection of roles or groups that need to be set</param>
        ''' <param name="GroupsOrRoles">The DataRowCollection that represents either roles or groups</param>
        ''' <param name="ColumnName">The column name to retrieve the data from</param>
        Private Shared Sub setRolesOrGroups(ByRef stringCollectionObject As Collection(Of String), ByVal groupsOrRoles As DataRowCollection, ByVal columnName As String)
            For Each mRow In groupsOrRoles
                If Not IsDBNull(mRow(columnName)) Then
                    stringCollectionObject.Add(mRow(columnName).ToString())
                End If
            Next
        End Sub

        ''' <summary>
        ''' Sets the assigned roles or groups.
        ''' </summary>
        ''' <param name="stringCollectionObject">The collection of roles or groups that need to be set</param>
        ''' <param name="commaSeparatedString">A comma Separated list of roles or groups 'you, me' as an example</param>
        Private Shared Sub setRolesOrGroups(ByRef stringCollectionObject As Collection(Of String), ByRef commaSeparatedString As String)
            Dim mRoles() As String = commaSeparatedString.Split(",")
            stringCollectionObject = New Collection(Of String)
            For Each mRole In mRoles
                stringCollectionObject.Add(mRole.ToString())
            Next
        End Sub

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="collectionOfStrings"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Private Shared Function getCommaSeportatedString(ByVal collectionOfStrings As Collection(Of String)) As String
            Dim mRetValue As String = String.Empty
            If Not collectionOfStrings Is Nothing Then
                If collectionOfStrings.Count > 0 Then
                    For Each item In collectionOfStrings
                        mRetValue += item.ToString() + ","
                    Next
                End If
            End If
            If mRetValue.Length > 0 Then
                mRetValue = mRetValue.Substring(0, mRetValue.Length - 1)
            End If
            Return mRetValue
        End Function

#End Region
    End Class
End Namespace
